### PR TITLE
GH-2741 queryevaluation precompile

### PR DESCRIPTION
**Pull request by JervenBolleman**
GitHub issue resolved: #2741 

Briefly describe the changes proposed in this PR:

This is a clean series of commits that implement a query pre-compilation step. For each phase of the query evaluation we attempt to resolve all "constants for the life time of the query" and use them in java lamba's to do the actual evaluation.

It does so by introducing query evaluation steps. These are constructed once for the query and do the actual evaluation.
This logic allows us to reuse objects like lamba's which are relatively expensive to construct per invocation but should only be created once per query evaluation. For example we can now construct a near space optimal ArrayBindingSet, where we use lamba's to change lookups into direct index based array accesses.

This is the foundational feature with more improvements coming after it.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by kenwenzel** [comment]
Thank you for the good work. I've just done two short benchmark runs with your proposed changes.

```
# lmdb-lazy + GH-2741-queryevaluation-precompile
Benchmark                                               Mode  Cnt     Score     Error  Units
QueryBenchmark.complexQuery                             avgt    5    28.191 ±   4.948  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   379.815 ±  26.197  ms/op
QueryBenchmark.groupByQuery                             avgt    5    11.154 ±   0.697  ms/op
QueryBenchmark.removeByQuery                            avgt    5   263.245 ±  36.484  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5   661.805 ±  79.732  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5   494.132 ± 117.713  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  1047.549 ± 262.049  ms/op
```

```
# lmdb-lazy
Benchmark                                               Mode  Cnt     Score     Error  Units
QueryBenchmark.complexQuery                             avgt    5    29.223 ±   4.899  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   454.409 ±  62.850  ms/op
QueryBenchmark.groupByQuery                             avgt    5     6.847 ±   0.668  ms/op
QueryBenchmark.removeByQuery                            avgt    5   352.492 ± 284.050  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5   765.650 ± 180.301  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5   484.997 ±  58.529  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  1172.697 ±  87.222  ms/op
```

Findings: Most benchmarks are faster but especially groupByQuery is much slower. 
The removeX benchmarks have a lower fluctuation in runtime while the fluctuation of runtime for the simpleUpdateX
benchmarks increased. Maybe this is due to cache cleanups or other GC operations?!

Note: The benchmarks are executed on a Surface Book 3 with Linux in a Virtual Box VM that runs on Windows with anti virus software 🤣 This may also be the source for the observed fluctuations in runtime.

**Comment by kenwenzel** [comment]
The "groupBy issue" seems to also persist for more complex queries.

I've used similar data and a comparable query that I've found within this conversation:
https://phabricator.wikimedia.org/T291903

@JervenBolleman Do you have any idea what could be the reason for slower group by evaluations?

https://github.com/kenwenzel/rdf4j/tree/complex-benchmark-GH-2741
```
# lmdb-lazy + GH-2741-queryevaluation-precompile
Benchmark                                 Mode  Cnt      Score      Error  Units
QueryBenchmarkComplex.queryPersons        avgt    5  12267.927 ± 2337.873  ms/op
QueryBenchmarkComplex.queryPersonsSorted  avgt    5  13664.544 ± 3806.117  ms/op
```

https://github.com/kenwenzel/rdf4j/tree/complex-benchmark
```
# lmdb-lazy
Benchmark                                 Mode  Cnt     Score      Error  Units
QueryBenchmarkComplex.queryPersons        avgt    5  5992.694 ±  143.885  ms/op
QueryBenchmarkComplex.queryPersonsSorted  avgt    5  6823.772 ± 2044.730  ms/op
```

**Comment by kenwenzel** [comment]
Your new implementation shows 2 dominant execution paths in VisualVM:
image

The "old" implementation has only one execution path:
image

Have you changed anything within the stream implementation?

**Comment by JervenBolleman** [comment]
@kenwenzel yes. I have a good idea why the group by iterators slow down. The commit for that is by itself quite large and I haven't fully tested it. But I think I will add it to this one. Basically in group by we ended up building a lot of lamba's per final binding set, this is a lot of overhead.

So a basic query like
```sparql
SELECT COUNT(?s) WHERE {?s ?p ?o}
```
was sped up.
but a query like
```sparql
SELECT ?o COUNT(?s) WHERE {?s ?p ?o} GROUP BY ?o
```
was slowed down a lot.

**Comment by hmottestad** [comment]
We should try some queries with a lot of effective joins, since the usual procedure for a join operation is to create one iterator for the left side and for each returned element create a new iterator binding that element. So I'm wondering if this will end up calling the precompile stage for every new iterator it retrieves.

**Comment by kenwenzel** [comment]
Could it be also the case that the query is evaluated twice for some reason? The two execution paths shown by VisualVM look suspicious.
I will check this tomorrow.

**Comment by hmottestad** [comment]
### develop branch
Screenshot 2021-12-16 at 20 04 30

### this branch
Screenshot 2021-12-16 at 20 04 12

**Comment by hmottestad** [comment]
There is a lot of work being done in the `evaluate()` method with this branch while originally the `evaluate()` method does almost nothing. So maybe it is effectively running the query twice.

**Comment by kenwenzel** [comment]
The GroupIterator is created twice by two threads:

Here:
image

And also here:
image

Have you added some sort of parallel execution on the streams?

**Comment by JervenBolleman** [comment]
No there should not be multiple threads running. I will double check that I
did not break something in the rebase.

**Comment by hmottestad** [comment]
It's not two threads. You can see that one of the stack traces shows GroupIterator <init>, which is when a `new` GroupIterator is being instantiated. This iterator is then being used again in the stack trace that starts with the stream `count()` call.

**Comment by JervenBolleman** [comment]
> We should try some queries with a lot of effective joins, since the usual procedure for a join operation is to create one iterator for the left side and for each returned element create a new iterator binding that element. So I'm wondering if this will end up calling the precompile stage for every new iterator it retrieves.

No this should not be the case. On creation of the join evaluation step bot the left and right sides of the joins are prepared. This prepared step for the right is then invoked to generate the new iterators for each "joining".

See [JoinQueryEvaluationStep](core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/evaluationsteps/JoinQueryEvaluationStep.java) and the [JoinIterator](core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/JoinIterator.java) look for `rightIter = preparedRight.evaluate(leftIter.next());`

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/GroupIterator.java [hunk 1]
```diff
@@ -106,16 +104,17 @@ public GroupIterator(EvaluationStrategy strategy, Group group, BindingSet parent
...
+						.make();
 			} catch (IOException e) {
 				throw new QueryEvaluationException("could not initialize temp db", e);
 			}
-			this.db = DBMaker.newFileDB(tempFile).deleteFilesAfterClose().closeOnJvmShutdown().make();
 		} else {
-			this.tempFile = null;
 			this.db = null;
 		}
-		super.setIterator(createIterator());
```
Comment (hmottestad) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/GroupIterator.java]: This was the issue @JervenBolleman @kenwenzel

Comment (hmottestad) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/GroupIterator.java]: There was a call to `createIterator()` inside the constructor of the GroupedIterator which wasn't there before. When I removed it performance improves.

Comment (JervenBolleman) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/GroupIterator.java]: Ah @hmottestad that is definitely a bug that snuck in during rebasing.

**Comment by JervenBolleman** [comment]
Added a commit that should further reduce the cost of the GroupIterator. Will run some benchmarks here first.
Can I rebase this against develop and force push as on my desktop I am locally blocked on #3519 ? 
And cherry-picking that and then undoing it is a bit tedious?

**Comment by hmottestad** [comment]
> Added a commit that should further reduce the cost of the GroupIterator. Will run some benchmarks here first.
> Can I rebase this against develop and force push as on my desktop I am locally blocked on #3519 ?
> And cherry-picking that and then undoing it is a bit tedious?

Sure! Go ahead :)

**Comment by JervenBolleman** [comment]
Ok, I rebased locally. Running mvn clean install, and will then run the benchmarks. I won't force push before @kenwenzel says ok.

**Comment by kenwenzel** [comment]
> Ok, I rebased locally. Running mvn clean install, and will then run the benchmarks. I won't force push before @kenwenzel says ok.

Go ahead!

**Comment by JervenBolleman** [comment]
lmdb at 40a398944f8e9036062bb5961f08c72118371bb as proposed for this pull request on my desktop machine java-17 ubuntu, AMD Ryzen 3 3200G, 36 GB RAM.
```
Benchmark                                               Mode  Cnt     Score    Error  Units
QueryBenchmark.complexQuery                             avgt    5    30.685 ±  0.605  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   841.533 ± 12.603  ms/op
QueryBenchmark.groupByQuery                             avgt    5    13.947 ±  0.632  ms/op
QueryBenchmark.removeByQuery                            avgt    5   283.259 ±  1.915  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5   659.701 ± 35.504  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5   507.830 ±  7.767  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  1004.898 ± 52.089  ms/op
```
memory at the same
```
Benchmark                            Mode  Cnt   Score   Error  Units
QueryBenchmark.complexQuery          avgt   10  80.594 ± 0.295  ms/op
QueryBenchmark.groupByQuery          avgt   10   4.192 ± 0.022  ms/op
QueryBenchmark.pathExpressionQuery1  avgt   10  41.775 ± 0.177  ms/op
QueryBenchmark.pathExpressionQuery2  avgt   10   3.895 ± 0.161  ms/op
```

nativerdf
```
QueryBenchmark.complexQuery                             avgt    5    26.993 ±   0.685  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   985.184 ±  14.078  ms/op
QueryBenchmark.groupByQuery                             avgt    5    14.230 ±   0.259  ms/op
QueryBenchmark.removeByQuery                            avgt    5  1152.147 ±  15.979  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5  4238.761 ± 114.258  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5  4642.014 ± 154.681  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  4739.967 ± 156.664  ms/op
```

The develop branch at 1041f0575fbe1069f74b42ebce781a340ad9dcf7
lmdb
```
QueryBenchmark.complexQuery                             avgt    5   29.650 ±   0.805  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5  899.882 ±  11.203  ms/op
QueryBenchmark.groupByQuery                             avgt    5   13.832 ±   0.395  ms/op
QueryBenchmark.removeByQuery                            avgt    5  281.620 ±   6.691  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5  651.844 ±  23.065  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5  517.574 ±   3.513  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  968.515 ± 125.990  ms/op
```
memory
```
QueryBenchmark.complexQuery          avgt   10  76.140 ± 0.136  ms/op
QueryBenchmark.groupByQuery          avgt   10   5.111 ± 0.034  ms/op
QueryBenchmark.pathExpressionQuery1  avgt   10  41.503 ± 0.159  ms/op
QueryBenchmark.pathExpressionQuery2  avgt   10   3.821 ± 0.033  ms/op
```
native rdf
```
QueryBenchmark.complexQuery                             avgt    5    27.712 ±   0.755  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   954.477 ±  34.175  ms/op
QueryBenchmark.groupByQuery                             avgt    5    15.641 ±   0.502  ms/op
QueryBenchmark.removeByQuery                            avgt    5  1150.198 ±  16.041  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5  4326.869 ±  84.036  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5  4746.881 ± 198.947  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  4849.862 ± 247.159  ms/op
```

A funny thing about the complexquery benchmark on the memory store is that it very consistent but slower in performance on this branch then on develop. Yet, the first few executions on develop are 20% slower than it's steady state.
e.g. on develop
```
# Warmup Iteration   1: 104.427 ms/op
# Warmup Iteration   2: 101.325 ms/op
...
# Warmup Iteration   6: 102.898 ms/op
# Warmup Iteration   7: 93.523 ms/op
# Warmup Iteration   8: 78.381 ms/op
```
while
```
# Warmup Iteration   1: 87.263 ms/op
# Warmup Iteration   2: 79.956 ms/op
...
# Warmup Iteration   6: 77.816 ms/op
# Warmup Iteration   7: 79.400 ms/op
# Warmup Iteration   8: 80.852 ms/op

```
Still this is not so great so I do need to have a look at why things.

**Comment by hmottestad** [comment]
@JervenBolleman the complexQuery benchmark in the MemoryStore is highly bound by one of the joins in the query. I've created a cache which resolves this issue, you could give that a try since it will highlight performance changes with the pre-compilation to a much higher degree.

Caching implementation: https://github.com/eclipse/rdf4j/pull/3524

**Comment by JervenBolleman** [comment]
I noticed, that in this push I am not actually using the `ArrayBindingSet` at all. That means all the code is just being a bit difficult about getting to the same hashmap as before. I will just go and bang my head against my desk for a bit :(

**Comment by jeenbroekstra** [comment]
Just jumping in to say that I'm massively impressed by the progress here. I'm still not able to spend as much time on RDF4J as I'd like but it looks like I'm not really needed. Great stuff everyone!

**Comment by JervenBolleman** [comment]
lmdb at 02f1236ea78004fffdd1f2330d44d0f8205c6b85
```
QueryBenchmark.complexQuery                             avgt    5   31.763 ±   1.129  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5  687.850 ±  11.721  ms/op
QueryBenchmark.groupByQuery                             avgt    5   12.970 ±   0.877  ms/op
QueryBenchmark.removeByQuery                            avgt    5  279.232 ±   2.868  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5  647.433 ±  20.321  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5  504.342 ±   9.320  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  962.069 ± 112.605  ms/op
```
memory
```
QueryBenchmark.complexQuery          avgt   10  67.500 ± 0.445  ms/op
QueryBenchmark.groupByQuery          avgt   10   4.138 ± 0.025  ms/op
QueryBenchmark.pathExpressionQuery1  avgt   10  35.338 ± 0.153  ms/op
QueryBenchmark.pathExpressionQuery2  avgt   10   2.497 ± 0.017  ms/op
```
native
```
QueryBenchmark.complexQuery                             avgt    5    25.993 ±   0.536  ms/op
QueryBenchmark.distinctPredicatesQuery                  avgt    5   815.399 ±  21.933  ms/op
QueryBenchmark.groupByQuery                             avgt    5    15.059 ±   0.717  ms/op
QueryBenchmark.removeByQuery                            avgt    5  1129.202 ±  12.665  ms/op
QueryBenchmark.removeByQueryReadCommitted               avgt    5  4305.142 ± 137.832  ms/op
QueryBenchmark.simpleUpdateQueryIsolationNone           avgt    5  4663.887 ± 115.681  ms/op
QueryBenchmark.simpleUpdateQueryIsolationReadCommitted  avgt    5  4676.402 ± 169.647  ms/op
```

**Comment by kenwenzel** [comment]
Looks really good :-) Especially distinctPredicates is much faster.

This does not yet include my work in the lmdb-lazy branch?

**Comment by JervenBolleman** [comment]
@kenwenzel no it doesn't include your latest work. It's mostly due to more compact ArrayBindingSet over the QueryBindingSet. Still that can be improved but requires a clearing  up of #3530 

I need to profile execution now and probably redo the unbound test. Also QueryBindingSet is still used in quite a few places. I'll open an issue on that next week.

**Comment by JervenBolleman** [comment]
A few new commits. But this no longer really affects the speed of the queries benchmarked. Important is a single fix for the regex compiled pattern logic.

**Review by jeenbroekstra — COMMENTED** [review]
I'm not fully up to speed with everything that's been going on here, but to me this looks solid. I'm fine with you merging this at your discretion @JervenBolleman . You'll need to tackle a few merge conflicts though - possibly caused by the removal of the old federation sail.

**Comment by hmottestad** [comment]
We have some more query benchmarks if you want to see if there are any more differences. 

I would also appreciate if this PR was merged soon. I have some binding set improvements that I would like to contribute as a stop gap before you get index based binding sets working.

**Comment by JervenBolleman** [comment]
I will merge it Monday 10th of January 2022. I will fix the merge conflicts.

**Pull request merged**